core/system/src/context.c

```diff
@@ -0,0 +1,283 @@
+ * upwards when nested context switches return. */
+TContextPreviousState g_context_previous_states[UVISOR_CONTEXT_MAX_DEPTH];
+
+/** Stack pointer for the stack of previous contexts
+ *
+ * @internal */
+uint32_t g_context_p;
+
+/** Push the previous state to the state stack and updates the current state.
+ *
+ * @internal
+ *
+ * @warning This function trusts all the arguments that are passed to it. Input
+ * verification should be performed by the caller.
+ *
+ * @param context_type[in]  Type of context switch to perform
+ * @param src_id[in]        ID of the box we are switching context from
+ * @param src_sp[in]        Stack pointer of the box we are switching from */
+static void context_state_push(TContextSwitchType context_type, uint8_t src_id, uint32_t src_sp)
+{
+    /* Check that the state stack does not overflow. */
+    if (g_context_p >= UVISOR_CONTEXT_MAX_DEPTH) {
+        HALT_ERROR(SANITY_CHECK_FAILED, "Context state stack overflow");
+    }
+
+    /* Push the source box state to the state stack. */
+    g_context_previous_states[g_context_p].type = context_type;
+    g_context_previous_states[g_context_p].src_id = src_id;
+    g_context_previous_states[g_context_p].src_sp = src_sp;
+    ++g_context_p;
+
+    /* Update the current state of the source box. */
+    g_context_current_states[src_id].sp = src_sp;
+}
+
+/** Pop the previous state from the state stack.
+ *
+ * @internal
+ *
+ * @warning This function trusts all the arguments that are passed to it. Input
+ * verification should be performed by the caller.
+ *
+ * @returns the pointer to the previous box context state. */
+static TContextPreviousState * context_state_pop(void)
+{
+    /* Check that the state stack does not underflow. */
+    if (!g_context_p) {
+        HALT_ERROR(SANITY_CHECK_FAILED, "Context state stack underflow");
+    }
+
+    /* Pop the source box state and return it to the caller. */
+    --g_context_p;
+    return &g_context_previous_states[g_context_p];
+}
+
+/* Return a pointer to the previous box context state. */
+TContextPreviousState * context_state_previous(void)
+{
+    if (!g_context_p) {
+        /* There is no previous state */
+        return NULL;
+    } else {
+        return &g_context_previous_states[g_context_p - 1];
+    }
+}
+
+/* Forge a new exception stack frame and copy arguments from an old one. */
+uint32_t context_forge_exc_sf(uint32_t src_sp, uint8_t dst_id, uint32_t dst_fn, uint32_t dst_lr, uint32_t xpsr, int nargs)
+{
+    uint8_t src_id;
+    uint32_t dst_sp;
+    uint32_t exc_sf_alignment;
+
+    /* Source box: Gather information from the current state. */
+    src_id = g_active_box;
+
+    /* Destination box: Gather information from the current state. */
+    /* Note: Here we allow source and destination box IDs to be the same, as
+     * this function is used also for IRQs. */
+    if (src_id != dst_id) {
+        dst_sp = g_context_current_states[dst_id].sp;
```

Comment:
@AlessandroA Consider to swap the if-branches for better readability (and turn the != to ==)